pkg/server/api/agent/v1/service.go

```diff
@@ -88,13 +104,45 @@ func (s *Service) CreateJoinToken(ctx context.Context, req *agent.CreateJoinToke
 		},
 	})
 	if err != nil {
-		log.WithError(err).Error("Failed to register token")
-		return nil, status.Errorf(codes.Internal, "Failed to register token: %v", err)
+		log.WithError(err).Error("Failed to create token")
+		return nil, status.Errorf(codes.Internal, "Failed to create token: %v", err)
+	}
+
+	if req.AgentId != nil {
+		err = s.createJoinTokenRegistrationEntry(ctx, req.Token, req.AgentId)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "Failed to create registration entry: %v", err)
+		}
 	}
 
 	return &types.JoinToken{Value: result.JoinToken.Token, ExpiresAt: expiry}, nil
 }
 
+func (s *Service) createJoinTokenRegistrationEntry(ctx context.Context, token string, agentID *types.SPIFFEID) error {
+	parentID, err := spiffeid.New(s.td, "spire", "agent", "join_token", token)
+	if err != nil {
+		return fmt.Errorf("unable to generate parent SPIFFE ID: %v", err)
+	}
+	if agentID.TrustDomain != s.td {
```

Comment:
This validation should probably happen above when you parse the ID but before creating the join token...